sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMemoryLeaseManager.scala

```diff
@@ -0,0 +1,347 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids
+
+import ai.rapids.cudf.{Cuda, NvtxColor, NvtxRange, Rmm}
+import java.util
+import java.util.concurrent.ConcurrentHashMap
+
+import org.apache.spark.TaskContext
+import org.apache.spark.internal.Logging
+
+trait MemoryLease extends AutoCloseable {
+  /**
+   * Get the amount of memory assigned to this lease
+   */
+  def leaseAmount: Long
+}
+
+/**
+ * Allows tasks to coordinate between each other to avoid
+ * oversubscribing GPU memory.
+ */
+object GpuMemoryLeaseManager extends Logging {
+  // DO NOT ACCESS DIRECTLY!  Use `getInstance` instead.
+  @volatile private var instance: GpuMemoryLeaseManager = _
+
+  private def getInstance: GpuMemoryLeaseManager = {
+    if (instance == null) {
+      GpuMemoryLeaseManager.synchronized {
+        // The instance is trying to be used before it is initialized.
+        // Since we don't have access to a configuration object here,
+        // default to only one task per GPU behavior.
+        if (instance == null) {
+          initialize(0)
+        }
+      }
+    }
+    instance
+  }
+
+  /**
+   * Initializes the GPU memory lease manager
+   * @param savedForShuffle the amount of memory that is saved for suffle
+   */
+  def initialize(savedForShuffle: Long): Unit = synchronized {
```

Comment:
Do we want to call it reserved memory or something similarly more generic?  We may come up with other situations outside of shuffle that need memory set aside from this.  Or are you thinking we will add more parameters to this for each use-case that needs memory reserved in the future?

Along those same lines, I'm also wondering if it's this class's responsibility to calculate the memory amount to manage or the one calling it.  The benefit of having the caller specify the amount of memory to manage is that it makes it easier to unit test and reduces the coupling to other things (like shuffle memory reservation).

Reply:
If you want me to refactor it so we only have a pool size passed in I am fine with it.

Reply:
I do think it would be cleaner to move this out.  As it is now, the caller could ask for more memory to be reserved than there is memory available on the GPU, for example, which isn't covered.  By having the caller calculate the amount of memory for the lease manager to manage, I think we might be able to more cleanly handle these error cases and may be able to remove the need for standard vs. testing init.